http.go

```diff
@@ -251,6 +169,228 @@ func (g *Gateway) GraphQLHandler(w http.ResponseWriter, r *http.Request) {
+			payloadErr = err
+		}
+	}
+
+	// add the query to the list of operations
+	operations = append(operations, operation)
+
+	return
+}
+
+// Parses post request (plain or multipart) to list of operations
+func parsePostRequest(r *http.Request) (operations []*HTTPOperation, batchMode bool, payloadErr error) {
+	contentType := strings.SplitN(r.Header.Get("Content-Type"), ";", 2)[0]
+	switch contentType {
+	case "text/plain", "application/json", "":
+		// read the full request body
+		operationsJson, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			payloadErr = fmt.Errorf("encountered error reading body: %s", err.Error())
+			return
+		}
+
+		operations, batchMode, payloadErr = parseOperations(operationsJson)
+		break
+	case "multipart/form-data":
+
+		parseErr := r.ParseMultipartForm(32 << 20)
+		if parseErr != nil {
+			payloadErr = errors.New("error parse multipart request: " + parseErr.Error())
+			return
+		}
+
+		operationsJson := []byte(r.Form.Get("operations"))
+		operations, batchMode, payloadErr = parseOperations(operationsJson)
+
+		var filePosMap map[string][]string
+		if err := json.Unmarshal([]byte(r.Form.Get("map")), &filePosMap); err != nil {
+			payloadErr = errors.New("error parsing file map " + err.Error())
+			return
+		}
+
+		for filePos, paths := range filePosMap {
+			file, header, err := r.FormFile(filePos)
+			if err != nil {
+				payloadErr = errors.New("file with index not found: " + filePos)
+				return
+			}
+
+			fileMeta := graphql.Upload{
+				File:     file,
+				FileName: header.Filename,
+			}
+
+			if err := injectFile(operations, fileMeta, paths, batchMode); err != nil {
+				payloadErr = err
+				return
+			}
+		}
+		break
+	default:
+		payloadErr = errors.New("unknown content-type: " + contentType)
+		return
+	}
+
+	return
+}
+
+// Parses json operations string
+func parseOperations(operationsJson []byte) (operations []*HTTPOperation, batchMode bool, payloadErr error) {
+	// there are two possible options for receiving information from a post request
+	// the first is that the user provides an object in the form of { query, variables, operationName }
+	// the second option is a list of that object
+
+	singleQuery := &HTTPOperation{}
+	// if we were given a single object
+	if err := json.Unmarshal(operationsJson, &singleQuery); err == nil {
+		// add it to the list of operations
+		operations = append(operations, singleQuery)
+		// we weren't given an object
+	} else {
+		// but we could have been given a list
+		batch := []*HTTPOperation{}
+
+		if err = json.Unmarshal(operationsJson, &batch); err != nil {
+			payloadErr = fmt.Errorf("encountered error parsing operationsJson: %s", err.Error())
+		} else {
+			operations = batch
+		}
+
+		// we're in batch mode
+		batchMode = true
+	}
+
+	return operations, batchMode, payloadErr
+}
+
+// Adds file object to variables of respective operations in case of multipart request
```

Comment:
Can you explain a bit more what's going on  here? There is a lot of string parsing and I'm not sure I understand why

Reply:
Yeah, sure. In file upload request files are provided as separate "parts" of multipart request. They are connected to variables object through field called "map" (see https://github.com/obukhov/gateway/blob/support-file-upload/examples/fileupload/README.md for example). This function "injects" `Upload` type into `vartiables` object according to this `map` (see spec https://github.com/jaydenseric/graphql-multipart-request-spec) - map itself defines where to inject this file as a string (i.e. `variables.files.0` in map means that this file have to be attached to the variable called "files" and is a list of uploads to the position `0` in the list) - all string parsing is actually implementing these rules. This function also supports batch mode (where this map string will look like `1.variables.files.0` which makes it mapped to the second request in the batch) - this certainly adds some complexity here.

Reply:
Do you have an idea how to improve comment there to make it easier to understand? Maybe you can suggest better naming for the function. I guess a lot of string parsing there is because I don't trust "map" format and ensure that it is formed according to the spec in anyparticular detail.

Reply:
I think this might be one of those parts that's unavoidably complicated. I am not too bothered by the function's contents (the name is fine too). Once I read the spec (and your explanation) it was much more clear what was going on. Thank you for taking the time to walk me through all of this!

Reply:
I'm not too sure on the exact comment just yet but I'm sure something will come to me the next time I come back to this part of the codebase